streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -583,16 +583,19 @@ private void runLoop() {
                 }
             }
 
-            // try to process one fetch record from each task via the topology, and also trigger punctuate
+            // try to process several records from each task via the topology, and also trigger punctuate
             // functions if necessary, which may result in more records going through the topology in this loop
             if (totalNumBuffered > 0 || polledRecords) {
                 totalNumBuffered = 0;
 
                 if (!activeTasks.isEmpty()) {
                     for (StreamTask task : activeTasks.values()) {
-
-                        totalNumBuffered += task.process();
-
+                        // process this task's records to completion before
+                        // context switching to another task
+                        int numProcessed;
+                        while ((numProcessed = task.process()) != 0) {
```

Comment:
I'd propose to loop over the tasks with one record a time, i.e. if we have two tasks and each with 3 records, process them in one loop as:

```
poll -> task1, record1 -> task2, record1 -> task1, record2 -> task2, record2 -> task1, record3 -> task2, record3
```

I feel it is more "fair" and will not semi-starve some of the tasks if other task's traffic is much higher.